### ISSUE TITLE
inproc_thr multithread doesn't worked.

### ISSUE THREAD

**Issue by sailfish009** [nanomsg/nanomsg#791]
Hi, i modified following sample file,
we test it with h.264 hardware encoder.

nanomsg-1.0.0\perf\inproc_thr.c

worker thread and console program(main) works fine. (4 FULL HD file created, 1 frame : 3110400 bytes).
but if i modified following method, then it didn't worked. 

1. move bind socket part in main() functions to another custom function( bind_function(int socket)).
2. move nn_recv() in main() functions to another custom thread.
   tried parameter with std::ref(), but doesn't worked.

**Comment by gdamore** [comment]
Can you post your modified code, and the "test results" in more detail?  I need more to go on than this.

**Comment by sailfish009** [comment]
i can't copy our source code, 
but i can reproduce the similar code, here it is:
below code was works fine, but as i said before, 
if i move some part of code  to other function or other thread, then it didn't worked.

// working version
//////////////////////////////////////////////////////////////////////////
#define FRAME_SIZE  (3110400)      // YUV420 frame
#define FRAME_SIZE1 (2073600)      // Y  frame
#define FRAME_SIZE2 (518400)       // U, V  frame


DWORD WINAPI thread1(PVOID p_param)
{
  int rc;
  int s;

  s = nn_socket(AF_SP, NN_PAIR);
  assert(s != -1);
  rc = nn_connect(s, "inproc://inproc_thr");
  assert(rc >= 0);

  rc = nn_send(s, NULL, 0, 0);
  assert(rc == 0);

  while (1)
  {
    //event read
    //video data available
    ...
  
    unsigned char buf[1] = {channel_id};
    rc = nn_send(s, buf, 1, 0);
    assert(rc == 1);

    int size = 1920*1080;
    rc = nn_send(s, video_y, size, 0);
    assert(rc == size);
    int size2 = size / 4;
    rc = nn_send(s, video_u, size2, 0);
    assert(rc == size2);
    rc = nn_send(s, video_v, size2, 0);
    assert(rc == size2);
  }

  rc = nn_close(s);
  assert(rc == 0);

}

bool run()
{
  m_thread = CreateThread(NULL, 0x1000, thread1, this, 0, &tid);
  SetThreadPriority(m_thread, THREAD_PRIORITY_TIME_CRITICAL);

  return true;
}


int main()
{

  run();

  int socket = nn_socket (AF_SP, NN_PAIR);
  assert (socket != -1);
  rc = nn_bind (socket, "inproc://inproc_thr");
  assert (rc >= 0);

  int rc;
  unsigned char *buf_y;
  unsigned char *buf_u;
  unsigned char *buf_v;

  struct nn_stopwatch stopwatch;
  uint64_t elapsed;
  unsigned long throughput;
  double megabits;

  buf_y = (unsigned char*)malloc(FRAME_SIZE1);
  assert(buf_y);
  buf_u = (unsigned char*)malloc(FRAME_SIZE2);
  assert(buf_u);
  buf_v = (unsigned char*)malloc(FRAME_SIZE2);
  assert(buf_v);

  FILE * fp[4] = { nullptr };

  for (int i = 0; i < 4; ++i)
  {
    char str[128] = { 0 };
    sprintf(str, "test_%d.yuv", i);
    fp[i] = fopen(str, "wb");
  }

  unsigned char buf_id[1] = { 0 };
  rc = nn_recv(socket, buf_id, 1, 0);
  assert(rc == 0);

  nn_stopwatch_init(&stopwatch);
  for (int i = 0; i < 100; ++i)
  {
    rc = nn_recv(socket, buf_id, 1, 0);
    assert(rc == 1);

    rc = nn_recv(socket, buf_y, FRAME_SIZE1, 0);
    assert(rc == FRAME_SIZE1);

    rc = nn_recv(socket, buf_u, FRAME_SIZE2, 0);
    assert(rc == FRAME_SIZE2);

    rc = nn_recv(socket, buf_v, FRAME_SIZE2, 0);
    assert(rc == FRAME_SIZE2);

    switch (buf_id[0])
    {
    case 0:
    case 1:
    case 2:
    case 3:
      fwrite(buf_y, 1, FRAME_SIZE1, fp[buf_id[0]]);
      fwrite(buf_u, 1, FRAME_SIZE2, fp[buf_id[0]]);
      fwrite(buf_v, 1, FRAME_SIZE2, fp[buf_id[0]]);
      break;
    default:
      printf("wrong channel id: %d\n", buf_id[0]);
      break;
    }
  }
  elapsed = nn_stopwatch_term(&stopwatch);

  free(buf_y);
  free(buf_u);
  free(buf_v);
  rc = nn_close(socket);
  assert(rc == 0);

  if (elapsed == 0)
    elapsed = 1;
  throughput = (unsigned long)
    ((double)FRAME_SIZE / (double)elapsed * 1000000);
  megabits = (double)(throughput * FRAME_SIZE * 8) / 1000000;

  printf("message size: %d [B]\n", (int)FRAME_SIZE);
  printf("message count: %d\n", 100);
  printf("mean throughput: %d [msg/s]\n", (int)throughput);
  printf("mean throughput: %.3f [Mb/s]\n", (double)megabits);

  for (int i = 0; i < 4; ++i)
    fclose(fp[i]);

}

// non-working version
//////////////////////////////////////////////////////////////////////////

void video1(int socket)
{
  int rc;
  unsigned char *buf_y;
  unsigned char *buf_u;
  unsigned char *buf_v;

  struct nn_stopwatch stopwatch;
  uint64_t elapsed;
  unsigned long throughput;
  double megabits;

  buf_y = (unsigned char*)malloc(FRAME_SIZE1);
  assert(buf_y);
  buf_u = (unsigned char*)malloc(FRAME_SIZE2);
  assert(buf_u);
  buf_v = (unsigned char*)malloc(FRAME_SIZE2);
  assert(buf_v);

  FILE * fp[4] = { nullptr };

  for (int i = 0; i < 4; ++i)
  {
    char str[128] = { 0 };
    sprintf(str, "test_%d.yuv", i);
    fp[i] = fopen(str, "wb");
  }

  unsigned char buf_id[1] = { 0 };
  rc = nn_recv(socket, buf_id, 1, 0);
  assert(rc == 0);

  nn_stopwatch_init(&stopwatch);
  for (int i = 0; i < 100; ++i)
  {
    rc = nn_recv(socket, buf_id, 1, 0);
    assert(rc == 1);

    rc = nn_recv(socket, buf_y, FRAME_SIZE1, 0);
    assert(rc == FRAME_SIZE1);

    rc = nn_recv(socket, buf_u, FRAME_SIZE2, 0);
    assert(rc == FRAME_SIZE2);

    rc = nn_recv(socket, buf_v, FRAME_SIZE2, 0);
    assert(rc == FRAME_SIZE2);

    switch (buf_id[0])
    {
    case 0:
    case 1:
    case 2:
    case 3:
      fwrite(buf_y, 1, FRAME_SIZE1, fp[buf_id[0]]);
      fwrite(buf_u, 1, FRAME_SIZE2, fp[buf_id[0]]);
      fwrite(buf_v, 1, FRAME_SIZE2, fp[buf_id[0]]);
      break;
    default:
      printf("wrong channel id: %d\n", buf_id[0]);
      break;
    }
  }
  elapsed = nn_stopwatch_term(&stopwatch);

  free(buf_y);
  free(buf_u);
  free(buf_v);
  rc = nn_close(socket);
  assert(rc == 0);

  if (elapsed == 0)
    elapsed = 1;
  throughput = (unsigned long)
    ((double)FRAME_SIZE / (double)elapsed * 1000000);
  megabits = (double)(throughput * FRAME_SIZE * 8) / 1000000;

  printf("message size: %d [B]\n", (int)FRAME_SIZE);
  printf("message count: %d\n", 100);
  printf("mean throughput: %d [msg/s]\n", (int)throughput);
  printf("mean throughput: %.3f [Mb/s]\n", (double)megabits);

  for (int i = 0; i < 4; ++i)
    fclose(fp[i]);

}


BOOL bind_socket(int& socket)
{
  int rc;

  socket = nn_socket(AF_SP, NN_PAIR);
  assert(socket != -1);
  rc = nn_bind(socket, "inproc://inproc_thr");
  assert(rc >= 0);

  return TRUE;
}

int main()
{
  int s; 
  bind_socket(s);

  //device init and run
  ...

  std::thread raw_thread(video1, std::ref(s));
  raw_thread.detach();
  

  return 0;
}

**Comment by gdamore** [comment]
thanks i will take a look. note that inproc has some known race conditions. a rewrite of it is pending.  note also that pair does not guarantee delivery or ordering.  if the queue is backed up messages will be dropped.

**Comment by sailfish009** [comment]
My mistake.
reimplementation code works fine!
sorry for the waste your time.
